docs/design/core/metrics/Design.md

```diff
@@ -0,0 +1,233 @@
+## Concepts
+### Metric
+* A representation of data collected
+* Metric can be one of the following types: Counter, Gauge, Timer
+* Metric can be associated to a category. Some of the metric categories are Default, HttpClient, Streaming etc
+
+### MetricRegistry
+
+* A MetricRegistry represent an interface to store the collected metric data. It can hold different types of Metrics described above
+* MetricRegistry is generic and not tied to specific category (ApiCall, HttpClient etc) of metrics.
+* Each API has it own instance of the MetricRegistry. All metrics collected in the ApiCall lifecycle are stored in that instance.
+* A MetricRegistry can store other instances of same type. This can be used to store metrics for each Attempt in an Api Call.
+* [Interface prototype](prototype/MetricRegistry.java)
+
+### MetricPublisher
+
+* A MetricPublisher represent an interface to publish the collected metrics to a external source.
+* SDK provides implementations to publish metrics to services like [Amazon CloudWatch](https://aws.amazon.com/cloudwatch/), [Client Side Monitoring](https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/sdk-metrics.html) (also known as AWS SDK Metrics for Enterprise Support)
+* Customers can implement the interface and register the custom implementation to publish metrics to a platform not supported in the SDK.
+* MetricPublishers can have different behaviors in terms of list of metrics to publish, publishing frequency, configuration needed to publish etc.
+* Metrics can be explicitly published to the platform by calling publish() method. This can be useful in scenarios when the application fails 
+and customer wants to flush metrics before exiting the application.
+* [Interface prototype](prototype/MetricPublisher.java)
+
+### Reporting
+
+* Reporting is transferring the collected metrics to Publishers. 
+* To report metrics to a publisher, call the registerMetrics(MetricRegistry) method on the MetricPublisher.
+* There is no requirement for Publisher to publish the reported metrics immediately after calling this method.
```

Comment:
Will there be a way to publish metric reports separately for each request? This is so that I can tie the metrics logging with my parent unit of work. 

This is useful for instance when I am serving a request for my callers and as part of that I need to call an AWS API, I would like to log metrics for that particular call along with other metrics for the incoming request to my service. For this to work, I would need to be able to attach a publisher at the request level. This publisher will contain other custom metrics and at the end of my incoming request processing would flush the collected metrics to my custom destination (log file, some destination over Http).

Also if a publisher is configured at the request level, and another one is configured at the SDK level, only one should be invoked with the request level taking precedence.

I had mentioned more details in https://github.com/aws/aws-sdk-java-v2/issues/23#issuecomment-530060824